### DISCUSSION TITLE
How to move an object that is covered by another object

### DISCUSSION THREAD

**Discussion by Jiiiin** [fabricjs/fabric.js#7925] [Q&A] [answered]
I have an object covered by another object ,I tried to provide a selector for the user to select and then resize or move.
My original plan was to set every object selecable false on the canvas and then use setActiveObject to select it , but i failed. Can anyone help me.
I also try to use perPixelTargetFind,but  when the object is completely coverd, is's not work. I am not going to set preserveObjectStacking false.

https://codesandbox.io/s/zealous-clarke-bc5w1v?file=/src/App.tsx:274-296

**Comment by ShaMan123** [comment]
>My original plan was to set every object selecable false on the canvas and then use setActiveObject to select it , but i failed.

Sounds like the best way to do it.


> Can anyone help me. I also try to use perPixelTargetFind,but when the object is completely coverd, is's not work

`perPixelTargetFind` will not work - it's based on pixels... so won't fit.

**Reply by Jiiiin** [reply]
> My original plan was to set every object selecable false on the canvas and then use setActiveObject to select it , but i failed. 

But this solution doesn't seem to work for me, When I select the next layer that is covered and want to drag it on the canvas to move, it looks like a mousedown is triggered or something, feels like it's blocked by the layer above， and  then it gets unselected

**Reply by ShaMan123** [reply]
It is the most strange behavior I have seen.
Using a timeout works but for some reason clicking the button does not.
https://codesandbox.io/s/sweet-silence-5wv7n4?file=/src/App.tsx

**Reply by ShaMan123** [reply]
~~Solved~~

https://codesandbox.io/s/sweet-silence-5wv7n4?file=/src/App.tsx
https://jsfiddle.net/4g7emL2q/2/

~~It seems that the divs for selecting the rects where triggered somehow with the canvas....???~~ 😬🤕

Something disgusting happened. Anyways...

**Reply by ShaMan123** [reply]
Believe it or not this looks like an issue with React@18
Downgraded to 17 and it works.
https://codesandbox.io/s/fancy-flower-dmt5tj?file=/src/index.tsx

**Reply by Jiiiin** [reply]
https://user-images.githubusercontent.com/22162992/166942174-161fd56c-1b16-48a8-81ec-07c11afc87fb.mov

**Reply by Jiiiin** [reply]
hhh, it seems to be a demo React version problem, but the main problem I encountered is not this, you can watch this video, when I want to drag, if the initial click position  is in the yellow rect , it will be discard orange rect selection.

**Reply by ShaMan123** [reply]
Try using the `evented` flag with `selectable` as well

**Reply by Jiiiin** [reply]
`const rect1 = new fabric.Rect({
      width: 300,
      height: 300,
      fill: "orange",
      selectable: false
      // evented: false
    });
    const rect2 = new fabric.Rect({
      width: 300,
      height: 300,
      fill: "yellow",
      selectable: false,
      evented: false
    });`
Do you mean to set the upper object evented to be false, ? but I expect the upper object can also be selected to move.

**Reply by ShaMan123** [reply]
It depends on your logic. What I did is a POC

**Comment by ShaMan123** [comment, marked as answer]
Bug in React@18

**Reply by ShaMan123** [reply]
consider opening an issue there.

**Reply by Jiiiin** [reply]
Ok, I'll open an issue later for react 18 bugs.
